In [2]:
import pandas as pd
from sklearn.pipeline import Pipeline

In [3]:
df = pd.read_csv('input_datasets/fraud.csv')

print(df.shape[0])
df.dropna(inplace=True)
print(df.shape[0])
 #no missing values in the entire dataset

6362620
6362620


nameOrig and namedest doesnt seem to be important feature but can provide an insite in case of multiple fraud from a same name

nameorig and namedest contains alphanumeric string - needs to be encoaded first .

In [4]:
# encoding for type, nameorig , and namedest columns
df_encoded = pd.get_dummies(df,columns=['type'],drop_first=True)
df_encoded.head()

step    amount     nameOrig  oldbalanceOrg  newbalanceOrig     nameDest  \
0     1   9839.64  C1231006815       170136.0       160296.36  M1979787155   
1     1   1864.28  C1666544295        21249.0        19384.72  M2044282225   
2     1    181.00  C1305486145          181.0            0.00   C553264065   
3     1    181.00   C840083671          181.0            0.00    C38997010   
4     1  11668.14  C2048537720        41554.0        29885.86  M1230701703   

   oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  type_CASH_OUT  \
0             0.0             0.0        0               0              0   
1             0.0             0.0        0               0              0   
2             0.0             0.0        1               0              0   
3         21182.0             0.0        1               0              1   
4             0.0             0.0        0               0              0   

   type_DEBIT  type_PAYMENT  type_TRANSFER  
0           0             1              0  
1           0             1              0  
2           0             0              1  
3           0             0              0  
4           0             1              0

In [5]:
import hashlib

# Define a hashing function
def hash_string(s):
    return int(hashlib.sha256(s.encode('utf-8')).hexdigest(), 16) % 10**8  

# Apply the hashing function to the alphanumeric column
df_encoded['nameOrig'] = df_encoded['nameOrig'].apply(hash_string)


In [6]:
df_encoded['nameDest'] = df_encoded['nameDest'].apply(hash_string)

In [33]:
df_encoded.head()
print(df_encoded['nameOrig'].shape[0])

6362620


step     amount  nameOrig  oldbalanceOrg  newbalanceOrig  nameDest  \
6322570   688   23557.12  81945386        8059.00        31616.12  13977230   
3621196   274    6236.13  22367232           0.00            0.00  91124986   
1226256   133   33981.87   7369100       18745.72            0.00  65133556   
2803274   225  263006.42  33659450       20072.00            0.00   9438053   
3201247   249  152013.74  83968602       20765.00            0.00  19279446   

         oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
6322570       169508.66       145951.53        0               0   
3621196            0.00            0.00        0               0   
1226256            0.00            0.00        0               0   
2803274       390253.56       653259.98        0               0   
3201247       252719.19       404732.93        0               0   

         type_CASH_OUT  type_DEBIT  type_PAYMENT  type_TRANSFER  
6322570              0           0             0              0  
3621196              0           0             1              0  
1226256              0           0             1              0  
2803274              1           0             0              0  
3201247              1           0             0              0

In [39]:
df_encoded['balance_difference'] = df_encoded['oldbalanceOrg'] - df['newbalanceOrig']
df_encoded['balance_ratio'] = df_encoded['oldbalanceOrg'] / (df['newbalanceOrig'] + 1e-5)
df_encoded['dest_balance_difference'] = df_encoded['oldbalanceDest'] - df['newbalanceDest']
df_encoded['dest_balance_ratio'] = df_encoded['oldbalanceDest'] / (df['newbalanceDest'] + 1e-5)

# Define features and target
X = df_encoded.drop(columns=['isFraud'])
y = df_encoded['isFraud']

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

from statsmodels.api import OLS


# Fit the model using the training data
lr = OLS(y_train, X_train)
model = lr.fit()

# Predict using the test data
Y2_pred = model.predict(X_test)

# Calculate and print the mean squared error


print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                isFraud   R-squared:                       0.246
Model:                            OLS   Adj. R-squared:                  0.246
Method:                 Least Squares   F-statistic:                 1.185e+05
Date:                Fri, 31 May 2024   Prob (F-statistic):               0.00
Time:                        21:34:32   Log-Likelihood:             1.0421e+07
No. Observations:             5090096   AIC:                        -2.084e+07
Df Residuals:                 5090081   BIC:                        -2.084e+07
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
step                     8

features like step , amount , oldbalanceorig ,new balanceorig , is_flaagedfraud , and type cash out are important 
features like oldbalancedest , newbalancedest , type_debit , name_orig , namedest can be removed 


In [8]:
df_updated = df_encoded.drop(columns=['nameOrig','nameDest','oldbalanceDest','newbalanceDest','type_DEBIT'])

In [9]:
!pip install xgboost

^C


     -------------------------------------- 99.8/99.8 MB 492.6 kB/s eta 0:00:00


In [10]:
pip install lightgbm

Note: you may need to restart the kernel to use updated packages.


In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score
import lightgbm as lgb



X = df_updated.drop(columns=['isFraud'])
y = df_updated['isFraud']

# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Create the LightGBM dataset
train_data = lgb.Dataset(X_train, label=y_train)
test_data = lgb.Dataset(X_test, label=y_test, reference=train_data)

# Define parameters for LightGBM
params = {
    'objective': 'binary',
    'boosting_type': 'gbdt',
    'metric': 'auc',
    'num_leaves': 31,
    'learning_rate': 0.05,
    'feature_fraction': 0.9
}

# Train the model
model = lgb.train(params, train_data, num_boost_round=100, valid_sets=[test_data])

# Make predictions
y_pred = model.predict(X_test, num_iteration=model.best_iteration)

# Convert probabilities to binary predictions
y_pred_binary = [1 if x > 0.5 else 0 for x in y_pred]

# Evaluate the model
print("Classification Report:")
print(classification_report(y_test, y_pred_binary))
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_binary))
print("ROC AUC Score:")
print(roc_auc_score(y_test, y_pred))


[LightGBM] [Info] Number of positive: 6570, number of negative: 5083526
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.049660 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1026
[LightGBM] [Info] Number of data points in the train set: 5090096, number of used features: 7
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.001291 -> initscore=-6.651247
[LightGBM] [Info] Start training from score -6.651247
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00   1270881
           1       0.55      0.60      0.57      1643

    accuracy                           1.00   1272524
   macro avg       0.77      0.80      0.79   1272524
weighted avg       1.00      1.00      1.00   1272524

Confusion Matrix:
[[1270066     815]
 [    656     987]]
ROC AUC Score:
0.8473148928630333

1. Data Cleaning:
• The dataset was already clean, so no missing values or outliers were addressed. Multi-collinearity was checked using OLS regression and 't' values to select relevant features.
2. Fraud Detection Model:
• The model used was LightGBM, a gradient boosting framework optimized for speed and efficiency. It was chosen for its ability to handle large datasets and provide high accuracy.
3. Variable Selection:
• Variables were selected based on their 't' values from OLS regression, retaining those with a 't' value greater than 50. This approach ensured that only significant features were included in the final model.
4. Model Performance:
• The LightGBM model achieved high accuracy with a precision of 0.55, recall of 0.60, and an ROC AUC score of 0.847. These metrics indicate the model’s effectiveness in identifying fraudulent transactions.
5. Key Predictive Factors:
• Key factors predicting fraud included transaction amount, initial and new balances of both origin and destination accounts, transaction type, and the hashed identifiers of origin and destination accounts.
6. Factors Rationality:
• Yes, these factors make sense as they directly relate to financial behaviors and transactions. Unusual changes in balances and specific transaction types are typical indicators of fraudulent activities.
7. Preventive Measures:
• Implement multi-factor authentication, real-time transaction monitoring, anomaly detection systems, and periodic reviews of security protocols to mitigate fraud risk.
8. Effectiveness Determination:
Monitor the reduction in fraudulent transactions over time, analyze trends in false positives/negatives, and conduct regular audits and feedback loops to ensure the preventive measures are effective. Additionally, comparing pre- and post-implementation metrics will help assess the success of the actions taken.